src/python/pants/backend/project_info/tasks/export_dep_as_jar.py

```diff
@@ -193,33 +195,30 @@ def _zip_sources(target, location, suffix=".jar"):
                     )
         return f
 
-    def _dependencies_to_include_in_libraries(
-        self, t, modulizable_target_set, dependencies_needed_in_classpath
-    ):
-        """NB: We need to pass dependencies_needed_in_classpath here to make sure we're being strict_deps-aware
-        when computing the dependencies."""
-
-        dependencies_to_include = []
-        self.context.build_graph.walk_transitive_dependency_graph(
-            [direct_dep.address for direct_dep in t.dependencies],
-            # NB: Dependency graph between modulizable targets is represented with modules,
-            #     so we don't need to expand those branches of the dep graph.
-            predicate=lambda dep: (dep not in modulizable_target_set)
-            and (dep in dependencies_needed_in_classpath),
-            work=lambda dep: dependencies_to_include.append(dep),
-        )
-        return list(sorted(dependencies_to_include))
-
     def _extract_arguments_with_prefix_from_zinc_args(self, args, prefix):
         return [option[len(prefix) :] for option in args if option.startswith(prefix)]
 
+
+    def _compute_transitive_source_dependencies(self, target: Target, info_entry: List[str], modulizable_target_set: Set[Target]) -> List[str]:
+        if self._is_strict_deps(target):
+            return info_entry
+        else:
+            transitive_targets = set(info_entry)
+            self.context.build_graph.walk_transitive_dependency_graph(
+                addresses=[target.address],
+                predicate=lambda d: d in modulizable_target_set,
+                work=lambda d: transitive_targets.add(d.address.spec),
+            )
+            return list(transitive_targets)
```

Comment:
Warning: converting `set -> list` is non-deterministic across Pants runs. Do you care about that? This is never cached by V2, correct?

If you do care about stable ordered, use `pants.util.ordered_set.OrderedSet` or, even better, `FrozenOrderedSet`..

Reply:
Also, why return a `list`? Unless you need to mutate the result, it's safer to return `Tuple[str, ...]`.

Reply:
I think I'm going to go with accumulating partial results in an `OrderedSet` and freezing it when we're done. I'd appreciate it if you could take a look to the two functions again, this is very informative.

Reply:
> I think I'm going to go with accumulating partial results in an OrderedSet and freezing it when we're done.

This is a fantastic pattern. Sounds good!